packages/video_player/video_player_web/lib/video_player_web.dart

```diff
@@ -144,10 +157,453 @@ class VideoPlayerPlugin extends VideoPlayerPlatform {
+      getFrame(widget.element);
+    }
+  }
+}
+
+class _TextureVideoPlayerRenderer extends _VideoPlayerRenderer {
+  const _TextureVideoPlayerRenderer({required super.element, super.paused});
+
+  @override
+  State createState() => _TextureVideoPlayerRendererState();
+}
+
+class _TextureVideoPlayerRendererState
+    extends _VideoPlayerRendererState<_TextureVideoPlayerRenderer> {
+  @override
+  Future<void> capture() async {
+    if (!widget.paused || lastFrameTime != widget.element.currentTime) {
+      lastFrameTime = widget.element.currentTime;
+      try {
+        final web.ImageBitmap newSource =
+            await promiseToFuture<web.ImageBitmap>(
+                web.window.createImageBitmap(widget.element));
+        final ui.Image img = await ui_web.createImageFromImageBitmap(newSource);
+
+        if (mounted) {
+          setState(() {
+            image?.dispose();
+            source?.close();
+            image = img;
+            source = newSource;
+          });
+        }
+      } on web.DOMException catch (err) {
+        lastFrameTime = null;
+        if (err.name == 'InvalidStateError') {
+          // We don't have enough data yet, continue on
+        } else {
+          rethrow;
+        }
+      }
+    }
+  }
+
+  ui.Image? image;
+
+  @override
+  void dispose() {
+    super.dispose();
+    image?.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return LayoutBuilder(
+        builder: (BuildContext context, BoxConstraints constraints) {
+      // adjust video element size to match dimensions of frame so we capture the correct sized bitmap
+      final double dpr = MediaQuery.of(context).devicePixelRatio;
+
+      final double maxWidth = constraints.maxWidth * dpr;
+      final double maxHeight = constraints.maxHeight * dpr;
+      final double videoWidth = widget.element.videoWidth.toDouble();
+      final double videoHeight = widget.element.videoHeight.toDouble();
+
+      widget.element.width =
+          (videoWidth == 0 ? maxWidth : min(videoWidth, maxWidth)).ceil();
+      widget.element.height =
+          (videoHeight == 0 ? maxHeight : min(videoHeight, maxHeight)).ceil();
+      if (image != null) {
+        return RawImage(image: image);
+      } else {
+        return const ColoredBox(color: Colors.black);
+      }
+    });
+  }
+}
+
+class _CanvasVideoPlayerRenderer extends _VideoPlayerRenderer {
+  const _CanvasVideoPlayerRenderer({required super.element, super.paused});
+
+  @override
+  State createState() => _CanvasVideoPlayerRendererState();
+}
+
+class _CanvasVideoPlayerRendererState
+    extends _VideoPlayerRendererState<_CanvasVideoPlayerRenderer> {
+  @override
+  Future<void> capture() async {
+    lastFrameTime = widget.element.currentTime;
+    if (canvas != null) {
+      final web.CanvasRenderingContext2D context =
+          canvas!.getContext('2d')! as web.CanvasRenderingContext2D;
+
+      context.drawImage(widget.element, 0, 0, canvas!.width, canvas!.height);
```

Comment:
This

Reply:
@ditman worth some benchmarking, but other than the native element, this is probably the fastest path to get a video frame rendered in FF and Safari. It avoids the transfer to the worker, which the slow path in FF / Safari and avoids the GC churn caused by getting bitmaps every frame. 

This mode only exists to allow the player to only use a platform layer while playing and use a texture while not playing. Ideally, there wouldn't be a need for the canvas element because ideally we could swap from a texture straight to the video element when it starts playing. However, the challenge is that you can't move a video element in the DOM without resetting it, so if you try to move the video element into the platform layer when it is created, it will immediately stop playing. 

I haven't experimented with invisible platform layers yet, but it's possible that an alternative could be using an invisible platform layer for the video when it's not playing and then swapping it to be a visible layer later.  If the video element was not removed from the DOM during a swap from invisible to visible, there's a possibility it could work, but depends on whether Chrome would allow us to get the frames while the video was in the invisible layer and whether the element would get recreated by that swap.

Reply:
@ditman worth some benchmarking, but other than the native element, this is probably the fastest path to get a video frame rendered in FF and Safari. It avoids the transfer to the worker, which the slow path in FF / Safari and avoids the GC churn caused by getting bitmaps every frame. 

This mode only exists to allow the player to only use a platform layer while playing and use a texture while not playing. Ideally, there wouldn't be a need for the canvas element because ideally we could swap from a texture straight to the video element when it starts playing. However, the challenge is that you can't move a video element in the DOM without resetting it, so if you try to move the video element into the platform layer when it is created, it will immediately stop playing. 

I haven't experimented with invisible platform layers yet, but it's possible that an alternative could be using an invisible platform layer for the video when it's not playing and then swapping it to be a visible layer later.  If the video element was not removed from the DOM during a swap from invisible to visible, there's a possibility it could work, but depends on whether Chrome would allow us to get the frames while the video was in the invisible layer and whether the element would get recreated by that swap.